pkg/uploadproxy/uploadproxy.go

```diff
@@ -0,0 +1,266 @@
+	uploadProxyPrivateKey *rsa.PrivateKey
+
+	// Used to verify token came from our apiserver.
+	apiServerPublicKey *rsa.PublicKey
+
+	uploadServerClient *http.Client
+}
+
+func NewUploadProxy(bindAddress string, bindPort uint, client kubernetes.Interface) (UploadProxyServer, error) {
+	var err error
+	app := &uploadProxyApp{
+		bindAddress: bindAddress,
+		bindPort:    bindPort,
+		client:      client,
+	}
+	app.certsDirectory, err = ioutil.TempDir("", "certsdir")
+	if err != nil {
+		return nil, errors.Errorf("Unable to create certs temporary directory: %v\n", errors.WithStack(err))
+	}
+
+	// generate/retrieve RSA key used to decrypt tokens
+	err = app.generateKeys()
+	if err != nil {
+		return nil, errors.Errorf("Unable to generate and retrieve rsa keys: %v", errors.WithStack(err))
+	}
+
+	// retrieve RSA key used by apiserver to sign tokens
+	err = app.getSigningKey()
+	if err != nil {
+		return nil, errors.Errorf("Unable to retrieve apiserver signing key: %v", errors.WithStack(err))
+	}
+
+	// generate self signed cert
+	err = app.generateSelfSignedCert()
+	if err != nil {
+		return nil, errors.Errorf("Unable to create self signed certs for upload proxy: %v\n", errors.WithStack(err))
+
+	}
+
+	// get upload server http client
+	err = app.getUploadServerClient()
+	if err != nil {
+		return nil, errors.Errorf("Unable to create upload server client: %v\n", errors.WithStack(err))
+	}
+
+	return app, nil
+}
+
+func (app *uploadProxyApp) getUploadServerClient() error {
+	clientCert, err := tls.LoadX509KeyPair("/etc/tls/uploadproxy/tls.cert", "/etc/tls/uploadproxy/tls.key")
+	if err != nil {
+		return err
+	}
+
+	caCertBytes, err := ioutil.ReadFile("/etc/tls/uploadproxy/ca.cert")
+	if err != nil {
+		return err
+	}
+
+	caCertPool := x509.NewCertPool()
+	caCertPool.AppendCertsFromPEM(caCertBytes)
+
+	tlsConfig := &tls.Config{
+		Certificates: []tls.Certificate{clientCert},
+		RootCAs:      caCertPool,
+	}
+	tlsConfig.BuildNameToCertificate()
+
+	transport := &http.Transport{TLSClientConfig: tlsConfig}
+	client := &http.Client{Transport: transport}
+
+	app.uploadServerClient = client
+
+	return nil
+}
+
+func (app *uploadProxyApp) handleUploadRequest(w http.ResponseWriter, r *http.Request) {
+
+	encryptedTokenData := r.Header.Get("UPLOAD_TOKEN")
+	if encryptedTokenData == "" {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	tokenData, err := apiserver.DecryptToken(encryptedTokenData, app.uploadProxyPrivateKey, app.apiServerPublicKey)
+
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	glog.V(Vuser).Infof("Received valid token: pvc: %s, namespace: %s", tokenData.PvcName, tokenData.Namespace)
+
+	app.proxyUploadRequest(tokenData.Namespace, tokenData.PvcName, w, r)
+}
+
+func (app *uploadProxyApp) proxyUploadRequest(namespace, pvc string, w http.ResponseWriter, r *http.Request) {
+	url := fmt.Sprintf("https://%s.%s.svc%s", controller.GetUploadResourceName(pvc), namespace, uploadserver.GetUploadPath(pvc))
```

Comment:
I think you already mentioned this earlier. There's a race condition here between when the pod is actually available and when the upload request comes in.

One thought I had about all of this was that we could have the uploadproxy component have a Pod informer on it so it can learn of the upload server pods and detect both that the uploadserver is in a "ready" state and detect the IP of the uploadserver pod.

That would do two things for us.

1. Uses pod IP directly which means we don't have to mess with services or pod DNS entries.
2. Lets us avoid trying to forward requests to upload servers that aren't ready yet.

When an upload request comes in for a server that isn't ready yet, we could either attempt to block until the server comes online, or immediately return something like a '503' error indicating that the upload client should try again.

Reply:
For 1) we'd have to disable tls client hostname validation as ip is not known when cert is created.  

For 2) I don't think an informer it's worth it unless we do the blocking wait for pod to come up.  Without the informer, we could return a 503 if the dns name doesn't resolve.  I guess we could also block while the hostname doesn't resolve.

I think it would be great if the upload proxy and server were really dumb and didn't do any kubernetes stuff.

Reply:
Just had another idea.  I was going to modify the apiserver to only return UploadTokens for PVCs with the `cdi.kubevirt.io/storage.upload.target` annotation.  Can also make sure `cdi.kubevirt.io/storage.upload.pod.phase` is `Running`.

Reply:
> Just had another idea. I was going to modify the apiserver to only return UploadTokens for PVCs with the cdi.kubevirt.io/storage.upload.target annotation. Can also make sure cdi.kubevirt.io/storage.upload.pod.phase is Running.

Yeah, that could work. That ensures that a token will never be issued for a pod that is not ready. 

The "Running" phase wouldn't be quite enough in this case. We'd also have to verify that the container's http readiness checks have passed. The upload-controller could set that in an annotation as well which the apiserver could inspect.